### PR TITLE
 #354 - Support building on CentOS6 with Python2.6

### LINKED ISSUE #354: Python version under CentOS 6
When getting Open GEE to compile and execute under CentOS6, it was necessary to upgrade the installed version of python from 2.6 to 2.7.  This, however, broke other system functions that require python2.6, eg, 'yum'.  Python2.7 is required because it is included in Open GEE code (gepublishmanagerhelper.h, line 24 "#include <python2.7/Python.h>").

Can we remove the requirement of python2.7 from the code?  I didn't try 'alias'ing it.

### PR DESCRIPTION
**Pull request by tst-awankhede**
resolves  #354 
regular build and stage_install commands should work

In the http://localhost/admin there is a bug that some icons are not loading hence not showing the db name, but it is not related to this change , hence use geserveradmin commands to publish the db once pushed successfully with fusion.

I will track the icons bug separately.

### CODE REVIEW

**Comment by tst-awankhede** [comment]
updated CLA.

File: earth_enterprise/src/SConstruct [hunk 1]
```diff
@@ -367,6 +367,37 @@ installdirs = {
...
+  match = re.search('[0-9][0-9.]*', stdout)
+  # If match is None we don't know the version.
+  if match is None:
+    return None
+  version = match.group().split('.')
+
+  return version
+
+
+def get_cc_version(cc_path):
```
Comment (tst-ppenev) [inline earth_enterprise/src/SConstruct]: This looks like a duplicate of the function above, rather than a definition of the `is_version_ge` function.

File: earth_enterprise/src/server/wsgi/SConscript [hunk 2]
```diff
@@ -31,7 +31,7 @@ wsgi_env = env.Clone()
 
 root_dir = Dir(wsgi_env.exportdirs['root']).abspath
 python_bin = wsgi_env['python_bin']
-compiler_cmd = '%s -m compileall -f' % python_bin
+compiler_cmd = '%s -m py_compile ' % python_bin
```
Comment (tst-ppenev) [inline earth_enterprise/src/server/wsgi/SConscript]: Very minor: Looks like there's an extra space at the end of this command.

File: earth_enterprise/src/server/wsgi/serve/publish/publish_manager_helper.py [hunk 3]
```diff
@@ -1417,9 +1417,8 @@ def _GetVhNameToBaseUrlDict(self):
       (scheme://host[:port]) -  {vh_name: vh_base_url}
     """
     vh_list = self.QueryVhList()
-    vhname_to_baseurl_dct = {
-        vh_name: self.GetVhBaseUrl(vh_url, vh_ssl) for (
-            vh_name, vh_url, vh_ssl) in vh_list}
+    for (vh_name, vh_url, vh_ssl) in vh_list:
```
Comment (tst-ppenev) [inline earth_enterprise/src/server/wsgi/serve/publish/publish_manager_helper.py]: Q: Have you tested this function? Do you not need to define `vhname_to_base_url_dict = {}` before you can assign its elements?

**Comment by tst-lsavoie** [comment]
@tst-awankhede it looks like your email address wasn't set up correctly when you committed the changes.  It's showing up as invalid-email-address on the commit.  I think that's why it's complaining about CLAs.

File: earth_enterprise/src/SConstruct [hunk 4]
```diff
@@ -367,6 +367,37 @@ installdirs = {
     'htdocs': InstServerDir('gehttpd/htdocs'),
 }
 
+def get_cc_version(cc_path):
+  """Extracts the version number of a C/C++ compiler (like GCC) as a list of
+  strings."""
+
+  process = subprocess.Popen(
+    [cc_path, '--version'], stdout=subprocess.PIPE)
```
Comment (tst-lsavoie) [inline earth_enterprise/src/SConstruct]: Looks like this will need to be changed to `cc_dir`.

File: earth_enterprise/src/SConstruct [hunk 5]
```diff
@@ -367,6 +367,37 @@ installdirs = {
     'htdocs': InstServerDir('gehttpd/htdocs'),
 }
 
+def get_cc_version(cc_path):
+  """Extracts the version number of a C/C++ compiler (like GCC) as a list of
+  strings."""
+
+  process = subprocess.Popen(
```
Comment (tst-lsavoie) [inline earth_enterprise/src/SConstruct]: Minor: I know this is @tst-ppenev's code, but you could simplify this by using the check_output function.  It returns the output directly so you don't have to deal with redirecting the output and using process.communicate().

File: earth_enterprise/src/SConstruct [hunk 6]
```diff
@@ -367,6 +367,40 @@ installdirs = {
...
+  version = match.group().split('.')
+
+  return version
+
+def is_version_ge(version_components, comparand_components):
+  """Tests if a version number is greater than, or equal to another version
+  number.  Both version numbers are specified as lists of version components.
+  """
+
+  if version_components is None:
```
Comment (tst-lsavoie) [inline earth_enterprise/src/SConstruct]: This will fail if comparand_components is none.

**Comment by tst-awankhede** [comment]
@tst-lsavoie where is the email not setup ? It correctly resolves the author to tst-awankhede  profile. Also in git client, this is what is setup
screenshot from 2017-09-01 11-47-31

File: earth_enterprise/src/SConstruct (hunk 6 above)
Comment (tst-ppenev) [reply earth_enterprise/src/SConstruct]: Yes.  It also doesn't make much sense to try comparing a version to something that you don't specify.

Comment (tst-ppenev) [reply earth_enterprise/src/SConstruct]: I think that's fine.  A caller shouldn't be asking to compare a version to something the caller is not specified.

File: earth_enterprise/src/SConstruct [hunk 7]
```diff
@@ -367,6 +367,40 @@ installdirs = {
     'htdocs': InstServerDir('gehttpd/htdocs'),
 }
 
+def get_cc_version(cc_dir):
```
Comment (tst-ppenev) [inline earth_enterprise/src/SConstruct]: This should be changed back to `cc_path`. This is the path to the compiler whose version is being tested; it's not a path to a directory.

File: earth_enterprise/src/fusion/gepublish/gepublishmanagerhelper.h [hunk 8]
```diff
@@ -20,12 +20,11 @@
...
 
-#include <stddef.h>
+#include <Python.h>
+#include <cstddef>
 
-#include <python2.7/Python.h>  // Note: should be first in include list.
-
-#include <string>
-#include <vector>
+#include "string"
```
Comment (tst-ppenev) [inline earth_enterprise/src/fusion/gepublish/gepublishmanagerhelper.h]: Q: Why is this changed to `"string"` and `"vector"`? Aren't we including the standard library headers here?

File: earth_enterprise/src/server/wsgi/serve/publish/publish_manager_helper.py [hunk 9]
```diff
@@ -1417,9 +1417,9 @@ def _GetVhNameToBaseUrlDict(self):
       (scheme://host[:port]) -  {vh_name: vh_base_url}
     """
     vh_list = self.QueryVhList()
-    vhname_to_baseurl_dct = {
-        vh_name: self.GetVhBaseUrl(vh_url, vh_ssl) for (
-            vh_name, vh_url, vh_ssl) in vh_list}
+    vhname_to_baseurl_dct = []
```
Comment (tst-ppenev) [inline earth_enterprise/src/server/wsgi/serve/publish/publish_manager_helper.py]: Q: I thought this was supposed to be a dictionary, rather than a list. Are the `vh_name`s consecutive integers?

File: earth_enterprise/src/SConstruct (hunk 5 above)
Comment (tst-ppenev) [reply earth_enterprise/src/SConstruct]: I don't think that's available in Python 2.6 on RHEL 6.

**Comment by tst-lsavoie** [comment]
@tst-awankhede my only guess is that maybe you were working on a different machine or a VM or something like that.  Your other two commits came through fine.  In any case, we know the commit came from you, so I don't know that we need to worry about it much.

**Comment by tst-ccamp** [comment]
@asmitaw If you click on the commits tab it should show you that one of the commits has an incorrect email. 

image

When I look at the commit locally, I see:
```
commit d0d9fe138e4e4affc279e103abc073d1a157b4ca
Author: root <root@localhost.localdomain>
Date:   Wed Aug 30 10:36:44 2017 -0700

     #354 - Support building on CentOS6 with Python2.6
```

File: earth_enterprise/src/SConstruct (hunk 4 above)
Comment (tst-awankhede) [reply earth_enterprise/src/SConstruct]: its the compiler path, so leaving it as cc_path

File: earth_enterprise/src/installer/install_fusion.sh [hunk 10]
```diff
@@ -250,7 +250,7 @@ check_prereq_software()
 		check_prereq_software_retval=1
 	fi
 
-	if ! software_check "$script_name" "python2.7" "python-2.7.*" "python27-*"; then
```
Comment (tst-ppenev) [inline earth_enterprise/src/installer/install_fusion.sh]: I think instead of separating Cent OS from RHEL, and using it as a separation between Cent OS / RHEL 7 and Cent OS / RHEL 6 we should accept both Python 2.6 and Python 2.7 here, since GEE should now run with both.

We just need to change the `grep` expression to match both '2.6' and '2.7'.

File: earth_enterprise/src/installer/install_server.sh [hunk 11]
```diff
@@ -143,7 +143,7 @@ check_prereq_software()
     check_prereq_software_retval=1
   fi
 
-  if ! software_check "$script_name" "python2.7" "python-2.7.*" "python27-*"; then
```
Comment (tst-ppenev) [inline earth_enterprise/src/installer/install_server.sh]: Same comment about accepting both Python 2.6 and 2.7.

**Comment by tst-lsavoie** [comment]
When I try to build with these changes I get failures like the ones below:

```
cc1: error: -Werror=unused-result: No option -Wunused-result
cc1: error: -Werror=unused-local-typedefs: No option -Wunused-local-typedefs
```

I'm using the default compiler (gcc 4.4.7).  Do I need to update it to something newer?

**Comment by tst-ppenev** [comment]
@tst-lsavoie: That's what you get when you build with an old compiler. It doesn't support some of the warning command-line options.

**Comment by tst-lsavoie** [comment]
What's the best way to update the compiler on CentOS 6?  We're going to need instructions for this at some point - might as well get started now.

**Comment by tst-ccamp** [comment]
This has some instructions: https://gist.github.com/giwa/b1fb1e44dc0a7d270881

There's also a "cc_dir" flag where you can pass in the path to the compiler.

**Comment by tst-dsheremata** [comment]
@tst-lsavoie  - I've got some draft markdown as a start on how to setup the new compiler for RH.  It's not checked in yet, but coming in an MR soon. Here's a quick snippet: 

# GEE 5.2.1 Build Prerequisites (All platforms) 


* git 1.7.1+
* gcc 4.8.x
* scons 2.3.x+
* python 2.6.x+ 


# RHEL6 Build Notes

## GCC 4.8

To obtain GCC 4.8 on RHEL6, you must enable the DevToolSet repositories in the RHSM: 

```
sudo subscription-manager repos --enable=rhel-x86_64-workstation-dts-2

sudo yum install devtoolset-2-gcc devtoolset-2-binutils  devtoolset-2-gcc-c++ devtoolset-2-gcc-gfortran
```

The gcc 4.8 installation will be located in the `/opt/rh/devtoolset-2/root/usr/bin/` directory   To switch the environment to use gcc 4.8, run: 

```source /opt/rh/devtoolset-2/enable``` 

NOTE:  There is still an issue that we’re working where scons does NOT properly use the gcc 4.8 - we’re actively working to fix this. 


## SCONS 2.3.x+
Scons 2.0 provided by RHSM is not compatible, and it must be upgraded. A “monarch” package is provided from [scons.org](scons-2.5.1-1.noarch.rpm):

```
wget scons-2.5.1-1.noarch.rpm
sudo yum install scons-2.5.1-1.noarch.rpm
```

**Comment by tst-lsavoie** [comment]
@tst-dsheremata @tst-ccamp thanks.

**Comment by tst-lsavoie** [comment]
Is this supposed to support building portable as well or are we handling that separately?  I get python errors when building portable.

**Comment by tst-ppenev** [comment]
@tst-lsavoie: What errors are you getting for Portable? I was able to build and install, but I'm not using exactly this branch.

**Comment by tst-lsavoie** [comment]
```
  File "earthenterprise/earth_enterprise/src/portableserver/build.py", line 251, in <module>
    main(sys.argv)
  File "earthenterprise/earth_enterprise/src/portableserver/build.py", line 248, in main
    builder.build()
  File "earthenterprise/earth_enterprise/src/portableserver/build.py", line 93, in build
    self.build_fileunpacker()
  File "earthenterprise/earth_enterprise/src/portableserver/build.py", line 177, in build_fileunpacker
    build_and_test.main(['build_and_test.py', self.platform])
  File "/home/admin/earthenterprise/earth_enterprise/src/portableserver/build/fileunpacker-build/build_and_test.py", line 84, in main
    if BuildLibrary(os_dir, argv[1].lower()=="windows"):
  File "/home/admin/earthenterprise/earth_enterprise/src/portableserver/build/fileunpacker-build/build_and_test.py", line 46, in BuildLibrary
    result = util.ExecuteCmd(line.format(**build_vars), use_shell=True)
  File "/home/admin/earthenterprise/earth_enterprise/src/portableserver/build/fileunpacker-build/util.py", line 48, in ExecuteCmd
    print "Executing: {}".format(os_cmd if use_shell else ' '.join(os_cmd))
ValueError: zero length field name in format
```

It looks like the fix is similar to some of the other changes already made on this PR but this particular file wasn't modified.

**Comment by tst-ppenev** [comment]
Yeah. I thought that was fixed in this PR, but it may have been missed.

This is just the error message for not recognizing your OS that's failing to be displayed.

We can fix it in this pull request, or leave it for another.  I build and install Portable in my working branch.

**Comment by tst-ppenev** [comment]
@tst-dsheremata: I haven't had to upgrade `scons` on RHEL 6.

**Comment by tst-lsavoie** [comment]
I'm happy to leave it for another PR.  We just need to make sure we get everything working for both CentOS 6 and RHEL 6.

File: earth_enterprise/src/SConstruct (hunk 5 above)
Comment (tst-awankhede) [reply earth_enterprise/src/SConstruct]: sorry, that looks like out of scope for this ticket.

File: earth_enterprise/src/server/wsgi/serve/publish/publish_manager_helper.py (hunk 3 above)
Comment (tst-awankhede) [reply earth_enterprise/src/server/wsgi/serve/publish/publish_manager_helper.py]: those are virtual host names, I have not tested, but I corrected it as per your comments

File: earth_enterprise/src/SConstruct [hunk 12]
```diff
@@ -397,8 +395,6 @@ def is_version_ge(version_components, comparand_components):
       version_component = int(version_components[i])
     if int(comparand_components[i]) > version_component:
       return False
-    if int(comparand_components[i]) < version_component:
```
Comment (tst-ppenev) [inline earth_enterprise/src/SConstruct]: Q: Why was this removed? The comparison won't work correctly, if we don't make this comparison.

File: earth_enterprise/src/server/wsgi/serve/publish/publish_manager_helper.py (hunk 3 above)
Comment (tst-ppenev) [reply earth_enterprise/src/server/wsgi/serve/publish/publish_manager_helper.py]: Q: I don't think the original code created a dictionary mapping keys to lists of `self.GetVhBaseUrl(. . .)` results.  It seemed just a map: key -> `self.GetVhBaseUrl(. . .)`.

File: earth_enterprise/src/SConstruct (hunk 12 above)
Comment (tst-awankhede) [reply earth_enterprise/src/SConstruct]: why it would not work? I tested and it works. Basically, the line above was redundant, everything else than the check that is retained is and was suppose to return TRUE and its doing so.

File: earth_enterprise/src/server/wsgi/serve/publish/publish_manager_helper.py (hunk 3 above)
Comment (tst-awankhede) [reply earth_enterprise/src/server/wsgi/serve/publish/publish_manager_helper.py]: @tst-ppenev  sorry I am not sure I follow your comment.

File: earth_enterprise/src/SConstruct (hunk 12 above)
Comment (tst-ppenev) [reply earth_enterprise/src/SConstruct]: It will work for the specific case of comparing 4.4.7 with 4.8, but it won't work for comparing two versions in general.

The line is not redundant.

This is comparing two numbers digit by digit starting from the most significant digit.  At each step, you either know which number is bigger, or the two digits must be equal.  The following digits have no way of telling you which number is bigger, unless all the more significant ones are equal.

You can try this for yourself:

```Python
# p.is_version_ge() is the function with the less-than check removed.
# q.is_version_ge() is the function with the less-than check.

>>> p.is_version_ge([4, 4, 7], [4, 8])
False
>>> q.is_version_ge([4, 4, 7], [4, 8])
False
# Comparing 4.4.7 and 4.8 works with the buggy version.

>>> p.is_version_ge([5, 2, 1], [4, 3])
False
>>> q.is_version_ge([5, 2, 1], [4, 3])
True
# Comparing 5.2.1 and 4.3 doesn't.
```

File: earth_enterprise/src/server/wsgi/serve/publish/publish_manager_helper.py (hunk 3 above)
Comment (tst-ppenev) [reply earth_enterprise/src/server/wsgi/serve/publish/publish_manager_helper.py]: Try this for yourself.

File `p.py`:
```Python
class PublishManagerHelper(object):
  """Mock class with the _GetVhNameToBaseUrlDict() method on the
  'centos-6-support-new' branch."""

  def __init__(self, query_vh_list):
    self.query_vh_list = query_vh_list

  def QueryVhList(self):
    return self.query_vh_list

  def GetVhBaseUrl(self, vh_url, vh_ssl):
    return vh_url

  def _GetVhNameToBaseUrlDict(self):
    """Builds a dictionary mapping a virtual host name to a base URL.
    Returns:
      dictionary mapping a virtual host name to a base URL
      (scheme://host[:port]) -  {vh_name: vh_base_url}
    """
    vh_list = self.QueryVhList()
    vhname_to_baseurl_dct = dict(
        (vh_name,[]) for (vh_name, vh_url, vh_ssl) in vh_list)
    for (vh_name, vh_url, vh_ssl) in vh_list:
      vhname_to_baseurl_dct[vh_name].append(self.GetVhBaseUrl(vh_url, vh_ssl))

    return vhname_to_baseurl_dct
```

File `q.py`:
```Python
class PublishManagerHelper(object):
  """Mock class with the _GetVhNameToBaseUrlDict() method on the 'master'
  branch."""

  def __init__(self, query_vh_list):
    self.query_vh_list = query_vh_list

  def QueryVhList(self):
    return self.query_vh_list

  def GetVhBaseUrl(self, vh_url, vh_ssl):
    return vh_url

  def _GetVhNameToBaseUrlDict(self):
    """Builds a dictionary mapping a virtual host name to a base URL.
    Returns:
      dictionary mapping a virtual host name to a base URL
      (scheme://host[:port]) -  {vh_name: vh_base_url}
    """
    vh_list = self.QueryVhList()
    vhname_to_baseurl_dct = {
        vh_name: self.GetVhBaseUrl(vh_url, vh_ssl) for (
            vh_name, vh_url, vh_ssl) in vh_list}

    return vhname_to_baseurl_dct
```

Compare old method and updated method:
```Python
>>> import p
>>> import q
>>> u = p.PublishManagerHelper([('localhost', 'http://127.0.0.1', False), ('thanksgivinghost', 'https://bay.king.it', True)])
>>> u._GetVhNameToBaseUrlDict()
{'localhost': ['http://127.0.0.1'], 'thanksgivinghost': ['https://bay.king.it']}
# This method returns a dictionary with values which are lists.
>>> old = q.PublishManagerHelper([('localhost', 'http://127.0.0.1', False), ('thanksgivinghost', 'https://bay.king.it', True)])
>>> old._GetVhNameToBaseUrlDict()
{'localhost': 'http://127.0.0.1', 'thanksgivinghost': 'https://bay.king.it'}
# This method returns a dictionary with values which are not lists (strings in this case).
```

File: earth_enterprise/src/SConstruct (hunk 5 above)
Comment (tst-lsavoie) [reply earth_enterprise/src/SConstruct]: I'm OK with leaving it as is.

File: earth_enterprise/src/server/wsgi/serve/publish/publish_manager_helper.py [hunk 13]
```diff
@@ -1417,9 +1417,11 @@ def _GetVhNameToBaseUrlDict(self):
       (scheme://host[:port]) -  {vh_name: vh_base_url}
     """
     vh_list = self.QueryVhList()
-    vhname_to_baseurl_dct = {
-        vh_name: self.GetVhBaseUrl(vh_url, vh_ssl) for (
-            vh_name, vh_url, vh_ssl) in vh_list}
+    vhname_to_baseurl_dct = dict(
```
Comment (tst-ppenev) [inline earth_enterprise/src/server/wsgi/serve/publish/publish_manager_helper.py]: :stuck_out_tongue: You don't need a for loop to create an empty dictionary, but it will work.

It could be just:

```Python
    vh_list = self.QueryVhList()
    vhname_to_baseurl_dct = {}
  
    for (vh_name, vh_url, vh_ssl) in vh_list:
      vhname_to_baseurl_dct[vh_name] = self.GetVhBaseUrl(vh_url, vh_ssl)
    return vhname_to_baseurl_dct
```

Comment (tst-awankhede) [reply earth_enterprise/src/server/wsgi/serve/publish/publish_manager_helper.py]: vhname_to_baseurl_dct = {}   throws error in 2.6, hence this hack

Comment (tst-ppenev) [reply earth_enterprise/src/server/wsgi/serve/publish/publish_manager_helper.py]: I don't have a problem with it with Python 2.6.6.

Comment (tst-ppenev) [reply earth_enterprise/src/server/wsgi/serve/publish/publish_manager_helper.py]: You could also use `vhname_to_baseurl_dct = dict()`.

Comment (tst-awankhede) [reply earth_enterprise/src/server/wsgi/serve/publish/publish_manager_helper.py]: for first suggestion , as I mentioned this is the error:
>>> import p
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "p.py", line 22
    vhname_to_baseurl_dct = dict{}
                                ^
SyntaxError: invalid syntax

Comment (tst-awankhede) [reply earth_enterprise/src/server/wsgi/serve/publish/publish_manager_helper.py]: Since its approved, I am not planning to make any more changes although seemingly vhname_to_baseurl_dct = dict() this is not throwing error, but not tested, so not changing it.
